### ISSUE TITLE
Implement 1.58 bit BitDistiller

### ISSUE THREAD

**Issue by BrownianNotion** [BrownianNotion/BitDistiller#56]
## 1. Description

At the time BitDistiller was released, the 1.58 bit LLM had not been implemented. 1.58 bit provides a quantisation level between 2 bit and 1 bit. Implementing this is non-trivial as the quantisation/clipping functions of BitDistiller do not natively support 1.58 bit.

Ternary weight quantisation from reference [2]
```
def weight_quant(w):
    scale = 1.0 / w.abs().mean().clamp_(min=1e-5)
    u = (w * scale).round().clamp_(-1, 1) / scale
    return u
```

TODO list:
- [x] Adapt ternary quantisation for `qlinear` BitDistiller Layer (see hugging face link, can ignore the activation quant for now)
- [x] Understand and check if clipping/other architectural changes (see 1.58 bitnet paper and comment below) are needed.
- [x] Get training to run successfully on dry run
- [x] Check weight distribution looks somewhat close to 2 bit, but i didnt' check layer wise. 
- [x] Check if inference modifications needed
- [x] Add 1.58 implementation closer to BitDistiller, see 1.58-range-scale branch, calculates scale using range instead of mean like bitnet 1.58.
- [x] Compare implementations and eval on 2 epochs 
- [x] Full train run on the range scale, same hyperparams as BitDistiller
- [x] Eval and upload model

## 2. Hypothesis

Given the better performance of 1.58 bit vs 1 bit LLMs in pretraining, we should expect to also a significant boost in performance. If the architectural changes are small, we should expect 1.58 bit to perform between 1 and 2 bits in terms of benchmarks.

## 3. Results

[Closer to bitnet 1.58 implementation](https://huggingface.co/BrownianNotion/TinyLlama_v1.1_1.58bit_ternary)
[Closer to BitDistiller implementation](https://huggingface.co/BrownianNotion/TinyLlama_v1.1_1.58bit_ternary_range_scale) 

[Second implementation trained on 4 epochs](https://huggingface.co/BrownianNotion/TinyLlama_v1.1_1.58bit_ternary_range_scale_4epochs)

**Wandb run name:** TinyLlama_v1.1_1.58_ternary_2025-04-02 21:34:19.441882  

For model trained on 4 epochs, even thought the (eval) loss and PPL decreased, the metrics barely changed. No benchmark appears statistically different from 1 bit, which is pretty much as bad as the 1 bit. Not obvious what could be going wrong, haven't deep dived anything yet.

## References
1. https://arxiv.org/pdf/2402.17764
2. https://huggingface.co/blog/1_58_llm_extreme_quantization 

**Reviewer:** @fredericowieser

**Comment by BrownianNotion** [comment]
Based on rereading BitDistiller, I think this might be relatively simple - replace the quantisation with a ternary quantisation. Clipping should be independent, but I will need to double check this. The BitNet paper makes other architectural adjustments (eg. RMSNorm), but we can simply start with trying the 1.58 quantisation in order to fit it into the BitDistiller framework. This is still a novel contribution given that 1.58 bits didn't exist at the time.

**Comment by BrownianNotion** [comment]
Note to self: pseudo_quantise is repeated for some reason in `test_utils.py` for int quantisers, make sure to change this too during inference if needed and doing 1bit changse.

**Comment by BrownianNotion** [comment]
Neither method looks particularly promising atm (big drop from 2bit, might even be worse than 1 bit). But will probably run the range_scale method if any as there's not much difference in the benchmarks except on PPL where range scale far outshines, and is also closer to the BitDistiller implementation.
